consensus/executor/executor.go

```diff
@@ -0,0 +1,220 @@
+	co := &coordinatorImpl{
+		rawExecutor: rawExecutor,
+		consumer:    consumer,
+		stc:         statetransfer.NewCoordinatorImpl(stps),
+		manager:     events.NewManagerImpl(),
+	}
+	co.manager.SetReceiver(co)
+	return co
+}
+
+// ProcessEvent is the main event loop for the executor.Coordinator
+func (co *coordinatorImpl) ProcessEvent(event events.Event) events.Event {
+	switch et := event.(type) {
+	case executeEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			if logger.IsEnabledFor(logging.DEBUG) {
+				logger.Debug("Executor is processing an executeEvent")
+			}
+		}
+		if co.skipInProgress {
+			logger.Error("FATAL programming error, attempted to execute a transaction during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			if logger.IsEnabledFor(logging.DEBUG) {
+				logger.Debug("Starting new transaction batch")
+			}
+			co.batchInProgress = true
+			err := co.rawExecutor.BeginTxBatch(co)
+			_ = err // TODO This should probably panic, see issue 752
+		}
+
+		co.rawExecutor.ExecTxs(co, et.txs)
+
+		co.consumer.Executed(et.tag)
+	case commitEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			logger.Debug("Executor is processing an commitEvent")
+		}
+		if co.skipInProgress {
+			logger.Error("Likely FATAL programming error, attempted to commit a transaction batch during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			logger.Error("Likely FATAL programming error, attemted to commit a transaction batch when one does not exist")
+			return nil
+		}
+
+		_, err := co.rawExecutor.CommitTxBatch(co, et.metadata)
+		_ = err // TODO This should probably panic, see issue 752
+
+		co.batchInProgress = false
+
+		info := co.rawExecutor.GetBlockchainInfo()
+
+		co.consumer.Committed(et.tag, info)
+	case rollbackEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			logger.Debug("Executor is processing an rollbackEvent")
+		}
+		if co.skipInProgress {
+			logger.Error("Programming error, attempted to rollback a transaction batch during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			logger.Error("Programming error, attempted to rollback a transaction batch which had not started")
+			return nil
+		}
+
+		err := co.rawExecutor.RollbackTxBatch(co)
+		_ = err // TODO This should probably panic, see issue 752
+
+		co.batchInProgress = false
+
+		co.consumer.RolledBack(et.tag)
+	case stateUpdateEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			logger.Debug("Executor is processing an stateUpdateEvent")
+		}
+		if co.batchInProgress {
+			err := co.rawExecutor.RollbackTxBatch(co)
+			_ = err // TODO This should probably panic, see issue 752
+		}
+
+		co.skipInProgress = true
+
+		info := et.blockchainInfo
+		for {
+			err, recoverable := co.stc.SyncToTarget(info.Height-1, info.CurrentBlockHash, et.peers)
+			if err == nil {
+				co.skipInProgress = false
+				co.consumer.StateUpdated(et.tag, info)
+				return nil
+			}
+			if !recoverable {
+				if logger.IsEnabledFor(logging.DEBUG) {
```

Comment:
I considered making this a more severe message, but the `!recoverable` is less severe than it sounds.  It simply means that the state transfer target supplied is no longer reachable (usually because the target was sufficiently newer than our state that we could not recover via deltas, so we had to recover via a state snapshot, and that snapshot was for a block in the future, which we can't verify yet).